### DISCUSSION TITLE
Brotli compression of 9p filesystem

### DISCUSSION THREAD

**Discussion by krishenriksen** [copy/v86#776] [Ideas]
Managed to get `ubuntu-9p-rootfs-flat` based on Ubuntu 16.04 down to only 66.7 MB, while having state be 24.6 MB

This is due to compressing the 9p filesystem using Brotli, with the following changes in `browser/filestorage.js`

```JavaScript
ServerFileStorageWrapper.prototype.load_from_server = function(sha256sum)
{

    return new Promise((resolve, reject) =>
    {
        v86util.load_file(this.baseurl + sha256sum, { done: async buffer =>
        {
            let data = new Uint8Array(buffer);
            data = globalThis.BrotliDecode(data);
            await this.cache(sha256sum, data);
            resolve(data);
        }});
    });
};
```

globalThis.BrotliDecode is derived from the attached script.
[brotli_decode.js.zip](https://github.com/copy/v86/files/10148112/brotli_decode.js.zip)

The build.sh is as follows

```bash
#!/bin/bash

./build-container.sh
./build-state.js

# compress blockstate
zstd --ultra -22 -f ../../../images/blockstate.bin

# compress 9p filesystem
#zstd --ultra -22 -f ../../../images/ubuntu-9p-rootfs-flat/*.bin
brotli --best --verbose -f ../../../images/ubuntu-9p-rootfs-flat/*.bin
rm -rf ../../../images/ubuntu-9p-rootfs-flat/*.bin

# Rename all *.br to *.bin
for file in ../../../images/ubuntu-9p-rootfs-flat/*.br; do
  mv -- "$file" "${file%.bin.br}.bin"
done
```

**Comment by copy** [comment]
Note that http can transparently use brotli, e.g. https://github.com/google/ngx_brotli

If you pass the wasm object to the filesystem you could also use zstd. I did so recently for partfiles, and would also like to do this for the filesystem: https://github.com/copy/v86/commit/df1a7129f4e1a3877e8803614cf1d24769b87034
